File: backend/orchestrator/worktree.go
package orchestrator

import (
	"context"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strings"
)

// WorktreeManager handles git worktree operations.
type WorktreeManager struct {
	exec         Executor
	projectPath  string
	worktreeBase string
}

// NewWorktreeManager creates a new worktree manager.
func NewWorktreeManager(exec Executor, projectPath, worktreeBase string) *WorktreeManager {
	return &WorktreeManager{
		exec:         exec,
		projectPath:  projectPath,
		worktreeBase: worktreeBase,
	}
}

// CreateWorktree creates a new git worktree and branch for an issue.
// Returns the branch name and worktree path.
func (m *WorktreeManager) CreateWorktree(ctx context.Context, issueNumber int, issueTitle string) (branch, worktreePath string, err error) {
	// Generate branch name: issue-42-add-dark-mode
	branch = m.generateBranchName(issueNumber, issueTitle)

	// Determine worktree path
	worktreePath = filepath.Join(m.worktreeBase, branch)

	// Ensure worktree base directory exists
	if err := os.MkdirAll(m.worktreeBase, 0755); err != nil {
		return "", "", fmt.Errorf("failed to create worktree base directory: %w", err)
	}

	// Check if worktree already exists
	if _, err := os.Stat(worktreePath); err == nil {
		// Worktree exists, verify it's valid
		_, verifyErr := m.exec.RunDir(ctx, worktreePath, "git", "status")
		if verifyErr == nil {
			return branch, worktreePath, nil // Already exists and valid
		}
		// Invalid worktree, remove it
		if err := m.RemoveWorktree(ctx, worktreePath); err != nil {
			return "", "", fmt.Errorf("failed to clean up invalid worktree: %w", err)
		}
	}

	// Fetch latest from remote
	_, err = m.exec.RunDir(ctx, m.projectPath, "git", "fetch", "origin")
	if err != nil {
		return "", "", fmt.Errorf("failed to fetch from origin: %w", err)
	}

	// Check if branch exists on remote
	_, err = m.exec.RunDir(ctx, m.projectPath, "git", "rev-parse", "--verify", "origin/"+branch)
	branchExistsOnRemote := err == nil

	if branchExistsOnRemote {
		// Create worktree from existing remote branch
		_, err = m.exec.RunDir(ctx, m.projectPath, "git", "worktree", "add", worktreePath, "origin/"+branch)
		if err != nil {
			return "", "", fmt.Errorf("failed to create worktree from remote branch: %w", err)
		}
		// Set up tracking
		_, err = m.exec.RunDir(ctx, worktreePath, "git", "checkout", "-B", branch, "--track", "origin/"+branch)
		if err != nil {
			return "", "", fmt.Errorf("failed to set up tracking branch: %w", err)
		}
	} else {
		// Create new branch from main/master
		baseBranch := m.getDefaultBranch(ctx)
		_, err = m.exec.RunDir(ctx, m.projectPath, "git", "worktree", "add", "-b", branch, worktreePath, "origin/"+baseBranch)
		if err != nil {
			return "", "", fmt.Errorf("failed to create worktree with new branch: %w", err)
		}
	}

	return branch, worktreePath, nil
}

// RemoveWorktree removes a git worktree and optionally the branch.
func (m *WorktreeManager) RemoveWorktree(ctx context.Context, worktreePath string) error {
	// Force remove the worktree
	_, err := m.exec.RunDir(ctx, m.projectPath, "git", "worktree", "remove", "--force", worktreePath)
	if err != nil {
		// Try manual cleanup if git worktree remove fails
		if removeErr := os.RemoveAll(worktreePath); removeErr != nil {
			return fmt.Errorf("failed to remove worktree: %w", err)
		}
		// Prune worktree list
		m.exec.RunDir(ctx, m.projectPath, "git", "worktree", "prune")
	}
	return nil
}

// generateBranchName creates a branch name from issue number and title.
// Format: issue-42-add-dark-mode
func (m *WorktreeManager) generateBranchName(issueNumber int, issueTitle string) string {
	slug := slugify(issueTitle)
	return fmt.Sprintf("issue-%d-%s", issueNumber, slug)
}

// getDefaultBranch returns the default branch name (main or master).
func (m *WorktreeManager) getDefaultBranch(ctx context.Context) string {
	// Try main first
	_, err := m.exec.RunDir(ctx, m.projectPath, "git", "rev-parse", "--verify", "origin/main")
	if err == nil {
		return "main"
	}
	// Fall back to master
	return "master"
}

// slugify converts a string to a URL-friendly slug.
// "Add Dark Mode Support" -> "add-dark-mode-support"
func slugify(s string) string {
	// Convert to lowercase
	s = strings.ToLower(s)

	// Replace spaces and underscores with hyphens
	s = strings.ReplaceAll(s, " ", "-")
	s = strings.ReplaceAll(s, "_", "-")

	// Remove non-alphanumeric characters (except hyphens)
	reg := regexp.MustCompile(`[^a-z0-9-]`)
	s = reg.ReplaceAllString(s, "")

	// Replace multiple consecutive hyphens with single hyphen
	reg = regexp.MustCompile(`-+`)
	s = reg.ReplaceAllString(s, "-")

	// Trim leading/trailing hyphens
	s = strings.Trim(s, "-")

	// Limit length to avoid overly long branch names
	if len(s) > 50 {
		s = s[:50]
		// Don't end with a hyphen
		s = strings.TrimRight(s, "-")
	}

	return s
}

File: src/services/worktree.ts
/**
 * Git Worktree Service
 * Creates and manages git worktrees for parallel work sessions
 */

import { exists, remove } from "@tauri-apps/plugin-fs";
import type { WorktreeResult } from "@core/types/result";
import { ok, err, createWorktreeError } from "@core/types/result";
import { logWorktree } from "./logging";
import { executeGit, type CommandResult } from "./commandExecutor";

// ============================================================================
// Constants
// ============================================================================

/** Directory where worktrees are stored (relative to repo root) */
const WORKTREES_DIR = ".worktrees";

/** Default timeout for git commands (30 seconds) */
const DEFAULT_TIMEOUT_MS = 30000;

// ============================================================================
// Branch Name Generation
// ============================================================================

/**
 * Generate a branch name from issue number and title
 * Format: {issueNumber}-{title-slug}
 * Example: 42-fix-login-bug
 *
 * Falls back to "issue" if title produces empty slug
 */
export function generateBranchName(issueNumber: number, title: string): string {
  const slug = title
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, "-") // Replace non-alphanumeric with hyphens
    .replace(/^-+|-+$/g, "")     // Trim leading/trailing hyphens
    .slice(0, 50);               // Limit length

  // Use "issue" as fallback if slug is empty (e.g., title was all special chars)
  const finalSlug = slug || "issue";

  return `${issueNumber}-${finalSlug}`;
}

/**
 * Parse issue number from a branch name
 * Expects format: {issueNumber}-{anything}
 * Example: "42-fix-login-bug" -> 42
 *
 * Returns null if branch name doesn't match the expected pattern
 */
export function parseIssueNumberFromBranch(branchName: string): number | null {
  // Match pattern: digits followed by a hyphen at the start
  const match = branchName.match(/^(\d+)-/);
  if (!match) return null;

  const issueNumber = parseInt(match[1], 10);
  return isNaN(issueNumber) ? null : issueNumber;
}

/**
 * Get the worktree path for a given branch name
 */
export function getWorktreePath(repoRoot: string, branchName: string): string {
  return `${repoRoot}/${WORKTREES_DIR}/${branchName}`;
}

// ============================================================================
// Git Command Execution
// ============================================================================

/**
 * Map CommandResult to WorktreeResult, handling git-specific error codes
 */
function mapCommandResultToWorktreeResult(
  result: CommandResult,
  commandPreview: string
): WorktreeResult<string> {
  // Success
  if (result.exitCode === 0) {
    logWorktree("debug", "Git command succeeded", { command: commandPreview });
    return ok(result.stdout);
  }

  const stderr = result.stderr;

  // Handle timeout
  if (result.timedOut) {
    logWorktree("error", "Git command timed out", { command: commandPreview });
    return err(
      createWorktreeError(
        "worktree_create_failed",
        "Git command timed out",
        `${commandPreview} exceeded timeout`
      )
    );
  }

  // Handle branch already checked out
  if (stderr.includes("already checked out") || stderr.includes("is already used")) {
    logWorktree("error", "Branch already checked out", { command: commandPreview });
    return err(
      createWorktreeError(
        "branch_checked_out",
        "Branch is already checked out in another worktree",
        stderr.trim()
      )
    );
  }

  // Handle worktree already exists
  if (stderr.includes("already exists")) {
    logWorktree("error", "Worktree already exists", { command: commandPreview });
    return err(
      createWorktreeError(
        "worktree_exists",
        "Worktree already exists at this location",
        stderr.trim()
      )
    );
  }

  // Handle git not installed
  if (stderr.includes("ENOENT") || stderr.includes("not found")) {
    logWorktree("error", "Git not installed");
    return err(
      createWorktreeError(
        "git_not_installed",
        "Git is not installed",
        "Install git from https://git-scm.com"
      )
    );
  }

  // Generic command failure
  logWorktree("error", `Git command failed with exit code ${result.exitCode}`, {
    command: commandPreview,
    exitCode: result.exitCode,
    stderr: stderr.trim(),
  });
  return err(
    createWorktreeError(
      "worktree_create_failed",
      `Git command failed with exit code ${result.exitCode}`,
      stderr.trim()
    )
  );
}

/**
 * Execute a git command using the unified CommandExecutor
 */
async function execGit(
  args: string[],
  timeoutMs = DEFAULT_TIMEOUT_MS,
  suppressErrorLog = false
): Promise<WorktreeResult<string>> {
  const commandPreview = `git ${args.slice(0, 4).join(" ")}${args.length > 4 ? "..." : ""}`;
  logWorktree("debug", `Executing: ${commandPreview}`);

  const result = await executeGit(args, { timeoutMs, suppressErrorLog });

  if (result.ok) {
    return mapCommandResultToWorktreeResult(result.value, commandPreview);
  }

  // Map CommandError to WorktreeError
  const error = result.error;
  switch (error.type) {
    case "cancelled":
      return err(createWorktreeError("worktree_create_failed", "Operation cancelled"));
    case "not_installed":
      return err(
        createWorktreeError(
          "git_not_installed",
          "Git is not installed",
          "Install git from https://git-scm.com"
        )
      );
    default:
      return err(createWorktreeError("worktree_create_failed", error.message, error.details));
  }
}

// ============================================================================
// Branch Operations
// ============================================================================

/**
 * Check if a branch exists (locally or remotely)
 */
async function branchExists(branchName: string): Promise<boolean> {
  // suppressErrorLog: true because exit code 128 is expected when branch doesn't exist
  const result = await execGit(["rev-parse", "--verify", branchName], DEFAULT_TIMEOUT_MS, true);
  return result.ok;
}

/**
 * Check if a remote branch exists
 */
async function remoteBranchExists(branchName: string): Promise<boolean> {
  // suppressErrorLog: true because failure is expected when branch doesn't exist
  const result = await execGit(["ls-remote", "--heads", "origin", branchName], DEFAULT_TIMEOUT_MS, true);
  return result.ok && result.value.trim().length > 0;
}

// ============================================================================
// Worktree Operations
// ============================================================================

export interface WorktreeInfo {
  readonly path: string;
  readonly branchName: string;
}

/**
 * Create a git worktree for a branch
 * If the branch doesn't exist, creates it from the current HEAD
 */
export async function createWorktree(
  repoRoot: string,
  branchName: string,
  signal?: AbortSignal
): Promise<WorktreeResult<WorktreeInfo>> {
  const worktreePath = getWorktreePath(repoRoot, branchName);

  logWorktree("info", "Creating worktree", { branchName, path: worktreePath });

  // Check for cancellation
  if (signal?.aborted) {
    return err(createWorktreeError("worktree_create_failed", "Operation cancelled"));
  }

  // Check if worktree already exists in git - reuse it
  const gitKnowsWorktree = await worktreeExists(repoRoot, branchName);
  if (gitKnowsWorktree) {
    logWorktree("info", "Worktree already exists, reusing", { branchName, path: worktreePath });
    return ok({ path: worktreePath, branchName });
  }

  // Check if directory exists but git doesn't know about it (orphaned)
  const dirExists = await exists(worktreePath);
  if (dirExists) {
    logWorktree("warn", "Removing orphaned worktree directory", { branchName, path: worktreePath });
    try {
      await remove(worktreePath, { recursive: true });
    } catch (e) {
      return err(createWorktreeError(
        "worktree_create_failed",
        "Failed to remove orphaned worktree directory",
        String(e)
      ));
    }
  }

  // Check if branch exists
  const localBranchExists = await branchExists(branchName);
  const remoteExists = await remoteBranchExists(branchName);

  let result: WorktreeResult<string>;

  if (localBranchExists || remoteExists) {
    // Use existing branch
    logWorktree("debug", "Using existing branch", { branchName, local: localBranchExists, remote: remoteExists });
    result = await execGit(["worktree", "add", worktreePath, branchName]);
  } else {
    // Create new branch from current HEAD
    logWorktree("debug", "Creating new branch", { branchName });
    result = await execGit(["worktree", "add", "-b", branchName, worktreePath]);
  }

  if (!result.ok) {
    return result;
  }

  logWorktree("info", "Worktree created successfully", { branchName, path: worktreePath });
  return ok({ path: worktreePath, branchName });
}

/**
 * Remove a git worktree
 */
export async function removeWorktree(
  repoRoot: string,
  branchName: string
): Promise<WorktreeResult<void>> {
  const worktreePath = getWorktreePath(repoRoot, branchName);

  logWorktree("info", "Removing worktree", { branchName, path: worktreePath });

  // First try to remove the worktree
  const removeResult = await execGit(["worktree", "remove", worktreePath, "--force"]);

  if (!removeResult.ok) {
    logWorktree("error", "Failed to remove worktree", { branchName, error: removeResult.error.message });
    return err(
      createWorktreeError(
        "worktree_remove_failed",
        "Failed to remove worktree",
        removeResult.error.details
      )
    );
  }

  // Prune to clean up any stale references
  await execGit(["worktree", "prune"]);

  logWorktree("info", "Worktree removed successfully", { branchName });
  return ok(undefined);
}

/**
 * List all worktrees
 */
export async function listWorktrees(): Promise<WorktreeResult<WorktreeInfo[]>> {
  const result = await execGit(["worktree", "list", "--porcelain"]);

  if (!result.ok) {
    return result as WorktreeResult<WorktreeInfo[]>;
  }

  const worktrees: WorktreeInfo[] = [];
  const lines = result.value.split("\n");
  let currentPath = "";
  let currentBranch = "";

  for (const line of lines) {
    if (line.startsWith("worktree ")) {
      currentPath = line.slice(9);
    } else if (line.startsWith("branch refs/heads/")) {
      currentBranch = line.slice(18);
      if (currentPath && currentBranch) {
        worktrees.push({ path: currentPath, branchName: currentBranch });
      }
    } else if (line === "") {
      currentPath = "";
      currentBranch = "";
    }
  }

  return ok(worktrees);
}

/**
 * Check if a worktree exists for a given branch
 */
export async function worktreeExists(
  repoRoot: string,
  branchName: string
): Promise<boolean> {
  const result = await listWorktrees();
  if (!result.ok) {
    return false;
  }

  const targetPath = getWorktreePath(repoRoot, branchName);
  return result.value.some((wt) => wt.path === targetPath);
}
